### PR TITLE
gh-115999: Add free-threaded specialization for ``TO_BOOL``

### PR DESCRIPTION
**Pull request by corona10**
* None / bool / int / str are immutable types, so they is thread-safe.
* list is mutable, but by using ``PyList_GET_SIZE`` we can make it as thread-safe.

* Issue: gh-115999

### CODE REVIEW

File: Python/specialize.c
Comment (corona10) [inline Python/specialize.c]: @mpage I think that we should make here to be atomic right?
(at least verify that `cache->version <= version` with CAS)

File: Python/specialize.c
Comment (corona10) [inline Python/specialize.c]: There is a issue with CAS in macOS build, I am under investigation.

File: Python/specialize.c
Comment (eendebakpt) [inline Python/specialize.c:2690]: Maybe not for this PR to address, but why is the tuple type handled here? I see no specialization for tuple (or did I miss something?). Tuple is immutable, so it should be relatively easy to add

Comment (corona10) [reply Python/specialize.c:2690]: IIRC, we added specializaiton based on metric that is measured at https://github.com/faster-cpython/benchmarking-public.
I am not sure how many tuple cases are actually existed.

Comment (corona10) [reply Python/specialize.c:2690]: I agree that tuple case is easy to add but that would be based on how many boolean operations are existed for the tuple at the real world workload.

File: Python/specialize.c
Comment (mpage) [reply Python/specialize.c]: This doesn't need to be atomic. No other thread will modify or read this address. More generally, the only parts of the bytecode that may be accessed by other threads are the instructions (the opcode/oparg pair) and the adaptive counters. Both of these may be modified by another thread that is instrumenting the bytecode.

File: Python/specialize.c
Comment (mpage) [inline Python/specialize.c]: This is unnecessary. No other threads will be accessing the cached version.

File: Python/specialize.c
Comment (corona10) [reply Python/specialize.c]: Got it, I 've reverted the related logic.

File: Python/specialize.c
Comment (mpage) [inline Python/specialize.c]: This block of code needs to happen atomically with respect to other threads mutating the type. In free-threaded builds that requires holding the type lock. Structuring the code might be a little tricky, because I don't think we want to expose the type lock and we probably also don't want to move this logic into `typeobject.c`. One possible arrangement might be something like the following, where we keep the locking and version tag assignment in `typeobject.c` and the specialization logic in `specialize.c`:

```c
// Returns 0 on success or caller-specific error on failure.
typedef uint16_t (*_py_validate_type)(PyTypeObject *);

// Returns 0 on success, -1 if no type version could be assigned, or the error returned by validate
int
_PyType_Validate(PyTypeObject *ty, _py_validate_type validate, unsigned int *tp_version)
{
    int err;
    BEGIN_TYPE_LOCK();
    err = validate(ty);
    if (!err) {
        if(assign_version_tag(_PyInterpreterState_GET(), ty))
            *tp_version = ty->tp_version_tag;
        }
        else {
            err = -1;
        }
    }
    END_TYPE_LOCK();
    return err;
}

// The specialization code 

static uint16_t
check_type_always_true(PyTypeObject *ty)
{
    PyNumberMethods *nb = ty->tp_as_number;
    if (nb && nb->nb_bool) {
        return SPEC_FAIL_TO_BOOL_NUMBER;
    }
    PyMappingMethods *mp = Py_TYPE(value)->tp_as_mapping;
    if (mp && mp->mp_length) {
        return SPEC_FAIL_TO_BOOL_MAPPING;
    }
    PySequenceMethods *sq = Py_TYPE(value)->tp_as_sequence;
    if (sq && sq->sq_length) {
      return SPEC_FAIL_TO_BOOL_SEQUENCE;
    }
    return 0;
}


if (PyType_HasFeature(Py_TYPE(value), Py_TPFLAGS_HEAPTYPE)) {
    unsigned int version = 0;
    int err = _PyType_Validate(Py_TYPE(value), check_type_always_true, &version);
    if (err < 0) {
        unspecialize(instr, SPEC_FAIL_OUT_OF_VERSIONS);
    }
    elif (err > 0) {
        unspecialize(instr, err);
    }
    else {
        write_u32(cache->version, version);
        specialize(instr, TO_BOOL_ALWAYS_TRUE);
        assert(version);
    }
    return;
}
```

**Review by mpage — CHANGES_REQUESTED** [review]
Thanks for doing this! Looks good overall. In addition to the comment inline, we're also going to need to update `_GUARD_TYPE_VERSION` to load `tp->tp_version` tag using an `FT_ATOMIC_LOAD_UINT32_RELAXED`.

My review raced with your updates, I hope the comments still make sense.

File: Python/specialize.c
Comment (corona10) [reply Python/specialize.c]: Brilliant approach. Thank you for letting me know where I should care :)
I've addressed your code review!

**Review by Hustlersambitionz — COMMENTED** [review]
****

**Comment by corona10** [comment]
I have made the requested changes; please review again

File: Include/internal/pycore_typeobject.h
Comment (mpage) [inline Include/internal/pycore_typeobject.h]: Can you add a short high-level description of the function in addition to the return value documentation?

File: Python/specialize.c
Comment (mpage) [inline Python/specialize.c]: This feels like it might be a bit too defensive, but up to you. By construction,`_PyType_Validate` should never return 0 without setting a version. And if that invariant is violated for some reason then the `assert(version)` below should catch it.

**Review by mpage — APPROVED** [review]
Sorry for the delay in review here, was traveling for work. Looks good!

**Comment by corona10** [comment]
@mpage I've updated the PR please take a look again for comments:)

@markshannon  Would you like to take a look before merging this PR?

File: Include/internal/pycore_typeobject.h
Comment (mpage) [inline Include/internal/pycore_typeobject.h]: I don't think a new version will be assigned if the type already has a valid version

Comment (corona10) [reply Include/internal/pycore_typeobject.h]: Updated :)

File: Python/specialize.c
Comment (markshannon) [inline Python/specialize.c:2669]: This should be surrounded by `#ifdef Py_STATS`. We don't want to be calling this unless stats are enabled.

File: Python/specialize.c
Comment (markshannon) [inline Python/specialize.c]: Can we leave the gotos, and have one call to `specialize` at the end?
This seems very repetitive and adds the overhead of several calls.

File: Python/specialize.c
Comment (markshannon) [inline Python/specialize.c:2745]: This seems to be adding a fair bit of overhead and indirection.
Would it be sufficient to set the version number before performing the other checks, and check that it hasn't changed at the end? The extra check would be benign in the default build.

**Review by markshannon — CHANGES_REQUESTED** [review]
I'm a bit concerned about the overhead this adds for the default build.

Specialization failure kinds are only needed when `Py_STATS` is defined.

File: Python/specialize.c
Comment (corona10) [reply Python/specialize.c:2669]: @mpage 
I have same concerns with this change, this change was first introduced at https://github.com/python/cpython/pull/126414, do we need to make it as dummy function for the default build?

File: Python/specialize.c
Comment (mpage) [reply Python/specialize.c]: I don't think this is any more repetitive that what was previously here and is easier to reason about. It's also the same number of function calls (1) which will be inlined.

File: Python/specialize.c
Comment (mpage) [reply Python/specialize.c:2669]: I don't think the performance overhead of calling this will matter, but if it does, we could replace it with a dummy function that returns a special sentinel value indicating that stats are disabled.

File: Python/specialize.c
Comment (mpage) [reply Python/specialize.c:2745]: > This seems to be adding a fair bit of overhead and indirection.

This adds one extra function call (the call to `check_type_always_true`). 

> Would it be sufficient to set the version number before performing the other checks, and check that it hasn't changed at the end?

This might be possible, but would require using something like a sequence lock to do correctly, which is more complicated and harder to reason about than a solution that acquires the lock around the validation that needs to be performed.

**Comment by mpage** [comment]
@markshannon - [Performance](https://github.com/facebookexperimental/free-threading-benchmarking/tree/main/results/bm-20241113-3.14.0a1+-b1f1c71) is neutral on the default build.

File: Python/specialize.c
Comment (mpage) [reply Python/specialize.c:2669]: Both clang (18.18) and GCC (11.5.0) eliminate the call to `to_bool_fail_kind` if stats are disabled because the result of the call is unused.

**Comment by mpage** [comment]
@markshannon - I don't think we need to be concerned about the performance impact of this change on the default build:

1. [Results](https://github.com/facebookexperimental/free-threading-benchmarking/tree/main/results/bm-20241113-3.14.0a1+-b1f1c71) on pyperformance on the default build look neutral to slightly positive.
2. The call to `to_bool_fail_kind` is eliminated if stats are disabled because the result of the call is unused and the function is pure.
3. The change to call `_PyType_Validate` introduces only one layer of indirection: a function call to pre-existing code.
4. The generated code (built with GCC 11.5.0 and `--enable-optimizations`) for `_Py_Specialize_ToBool` is nearly identical between [main](https://gist.github.com/mpage/b85011c0fce048fcd07c04c7dcb74e88) and [this branch](https://gist.github.com/mpage/1d14fc96c8164906d9e4962d38c4a3e9). The generated code for the hot path is actually a bit smaller with this change, because GCC is able to outline the error path when we fail to assign a type version. The main overhead from this change is that we look up the generic opcode based on the instruction being specialized, rather than supplying it as a constant. This introduces an extra couple of loads when we have to set the opcode if specialization fails. If you are concerned about that we can pass the generic opcode to `specialize` in a separate PR. 

Please have another look.

**Comment by corona10** [comment]
I have made the requested changes; please review again

**Comment by corona10** [comment]
Mark, 
I have addressed your code review especially for https://github.com/python/cpython/pull/126616#discussion_r1838336161

I am okay with Surrounding `#ifdef Py_STATS`  but it will need extra PR for `specialize` and `unspecialize`  function, so I prefer to handle it separately, and I believe that @mpage will handle this.

Please take a look :)

File: Python/specialize.c
Comment (mpage) [reply Python/specialize.c]: To elaborate a bit, it is the same amount of repetition in either case. The only difference is what is being repeated. The previous approach repeats the assignment to the opcode and the goto:

```c
if (PyBool_Check(value)) {
    instr->op.code = TO_BOOL_BOOL;
    goto success;
}
```
while the new approach repeats the call to specialize and the return:
```c
if (PyBool_Check(value)) {
    specialize(instr, TO_BOOL_BOOL);
    return;
}
```

The benefit of the latter approach is that the control flow is easier to follow.

Finally, we make one call to `specialize` and that call is inlined. As noted below, the generated code for the two versions of `_Py_Specialize_ToBool` is nearly identical.

**Review by mpage — APPROVED** [review]
I still don't understand the rationale for keeping the gotos, but I care more about seeing this PR move forward. Accepting to unblock.

**Comment by corona10** [comment]
@mpage I 've followed the changes from https://github.com/python/cpython/pull/127030, Would you like to take a look again?

**Review by mpage — CHANGES_REQUESTED** [review]
LGTM!

**Review by mpage — APPROVED** [review]

**Comment by corona10** [comment]
I am merging this PR, @markshannon. Please let me know if we have to handle more things in the future PRs. Matt and I will try to adjust what you concerns :) cc @mpage

**Pull request merged**
